### PR TITLE
Add bundle writer abstraction

### PR DESCRIPTION
**Pull request by efritz**
Create an abstraction between the bundle manager database and the storage layer. This makes a solid interface that we can use to later abstract out the storage layer if we find something that works better.

This implements the symmetric part of https://github.com/sourcegraph/sourcegraph/pull/10208. This partially implements https://github.com/sourcegraph/sourcegraph/issues/9965.

### CODE REVIEW

File: ThirdPartyLicensesNpm.csv [hunk 1]
```diff
@@ -2,6 +2,18 @@
 "@babel/runtime@7.9.2","MIT","https://github.com/babel/babel"
 "@hot-loader/react-dom@16.13.0","MIT","https://github.com/hot-loader/react-dom"
 "@primer/octicons-react@9.6.0","MIT","https://github.com/primer/octicons"
+"@reach/accordion@0.9.1","MIT","https://github.com/reach/reach-ui"
```
Comment (efritz) [inline ThirdPartyLicensesNpm.csv:5]: I'm unsure why the changes in this PR required an update to this file when the dev checks ran? cc @sourcegraph/web

Comment (felixfbecker) [reply ThirdPartyLicensesNpm.csv:5]: It shouldn't - we don't require the license file to be kept up to date in CI (that would be very annoying with Renovate PRs...) 🤔 
Looking at the Buildkite logs, it seems like the `go-generate.sh` step failed (not the licenses check). Apparently it uses `git diff --exit-code`, but doesn't scope it properly to the right Go files, but instead includes the licenses CSV too

Comment (efritz) [reply ThirdPartyLicensesNpm.csv]: I don't think it's scoped improperly - go generate will not always generate go files.

Comment (felixfbecker) [reply ThirdPartyLicensesNpm.csv:5]: Not the `go generate`, but the `git diff --exit-code` 🙂 
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/check/go-generate.sh#L17:39
This should probably be scoped to only Go files or directories (not sure)

Comment (efritz) [reply ThirdPartyLicensesNpm.csv:5]: I know. If the output of the go generate is not a go file, then why would we want to scope the diff to go files only? We should ensure that any file generated as a result of a go generate is in git correctly.

We generate go files, but we also generate db schemas, docs, and markdown with go generate. These should be included in the diff.

My question is why did the licenses change with this set of changes at all?

Comment (felixfbecker) [reply ThirdPartyLicensesNpm.csv:5]: Because the license file is not required to be kept up to date. But CI does run the command to update it to check if any MPL deps were added, so that's why it's dirty in CI.

Is there a different way that CI could check only files actually generated by `go generate` did not change?

Comment (efritz) [reply ThirdPartyLicensesNpm.csv:5]: Ah I see the issue now.
